### PR TITLE
[multistage] [bugfix] Fix sending mailbox object leak

### PR DESCRIPTION
**Pull request by 61yao**
Fix sending mailbox object leak by moving the sending mailbox object into exchange, where the ownership of the object should lie.

This PR also changes the implementation of inMemoryMailbox a bit to be more aligned with GrpcMailbox.

### CODE REVIEW

**Comment by 61yao** [comment]
@xiangfu0

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java [hunk 1]
```diff
@@ -38,23 +39,25 @@ public abstract class BlockExchange {
   // TODO: Deduct this value via grpc config maximum byte size; and make it configurable with override.
   // TODO: Max block size is a soft limit. only counts fixedSize datatable byte buffer
   private static final int MAX_MAILBOX_CONTENT_SIZE_BYTES = 4 * 1024 * 1024;
-
-  private final MailboxService<TransferableBlock> _mailbox;
-  private final List<MailboxIdentifier> _destinations;
+  private final List<SendingMailbox> _sendingMailboxes;
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java:42]:
```suggestion
  private final List<SendingMailbox<TransferableBlock>> _sendingMailboxes;
```

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java [hunk 2]
```diff
@@ -63,52 +66,33 @@ public static BlockExchange getExchange(MailboxService<TransferableBlock> mailbo
...
-
+  protected void sendBlock(SendingMailbox sendingMailbox, TransferableBlock block) {
     if (block.isEndOfStreamBlock()) {
       sendingMailbox.send(block);
       sendingMailbox.complete();
       return;
     }
 
     DataBlock.Type type = block.getType();
+    // TODO: Move splitter into mailbox service.
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java:90]: why do we need to move splitter into mailbox service? did you mean mailbox operator?

**Review by walterddr — APPROVED** [review]
lgtm. thank you for the refactoring this is amazing!

one comment regarding generics. please fix and we should be good to merge

**Pull request merged**